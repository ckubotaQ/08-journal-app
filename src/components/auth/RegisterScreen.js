import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { removeErrorAction, setErrorAction } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';



export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const {msgError} = useSelector(state => state.ui);
    console.log(msgError)

    const [formValues, handleInputChange ] = useForm ({
        name: 'Carlos',
        email: 'ckubota54@gmail.com',
        password: '123456',
        password2: '123456'
    })
    const { name, email, password, password2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if(isFormValid()) {
            dispatch( startRegisterWithEmailPasswordName(email,password,name) )
        }
        
    }
    const isFormValid = () => {
        if(name.trim().length===0){
            dispatch( setErrorAction('name is Required'));
            return false;
        }else if( !validator.isEmail( email )){
            dispatch( setErrorAction('email is Required'));
            return false;
        }else if(password!==password2 && password.length<5) {
           dispatch( setErrorAction('password should be at least 6 characters and match each'));
            return false;
        }
        dispatch(removeErrorAction());
        return true;
    }
    
    return (
        <>
            <h1 className="auth__title">Register</h1>
            <form onSubmit={handleRegister}>
                { msgError && 
                (
                    <div
                className="auth__alert-error">
                    {msgError}
                </div>
                ) 
                }
                <input type="text" 
                placeholder="Name"
                name="name"
                className="auth__input"
                autoComplete="off"
                value={name}
                onChange={handleInputChange}
                />
                <input type="text" 
                placeholder="email"
                name="email"
                className="auth__input"
                autoComplete="off"
                value={email}
                onChange={handleInputChange}
                />
                <input type="password" 
                placeholder="password" 
                name="password"
                className="auth__input"
                value={password}  
                onChange={handleInputChange}
                />
                <input type="password" 
                placeholder="Confirm"
                name="password2"
                className="auth__input"
                value={password2}
                onChange={handleInputChange}
                />
                <button
                type="submit"
                className="btn btn-primary btn-block mb-5"
                >
                   Register
                </button>
                    
                    <Link 
                    to="/auth/login"
                    className="link"
                    >

                        Alredy registered?
                    
                    </Link>


            </form>
        </>
    )
}
