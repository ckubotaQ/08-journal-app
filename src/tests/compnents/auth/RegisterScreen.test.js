import configureStore from 'redux-mock-store' //ES6 modules
import thunk from "redux-thunk";
import { Provider } from "react-redux"
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import '../../../setupTest';
import { types } from '../../../types/types';
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore(initState);

const wrapper = mount( 
    <Provider store={ store }>
        <MemoryRouter>
            <RegisterScreen /> 
        </MemoryRouter>
    </Provider>

)
describe('Pruebas en <RegisterScreeb />', () => {

    test('Debe de mostarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })
    test('debe de hacer el dispatch de la accion respectiva ', () => {
        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change',{
            target: {
                value:'',
                name: 'email'
            }
        })
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });
        const action = store.getActions();
        expect(action[0]).toEqual({
            type: types.uiSetError, 
            payload: 'email is Required'
        })
    })
    test('debe de mostrar la caja de alerta con el error', () => {
        const initialState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email no es correcto'
            }
        }
        const store = mockStore(initialState);
        const wrapper = mount( 
            <Provider store={ store }>
                <MemoryRouter>
                    <RegisterScreen /> 
                </MemoryRouter>
            </Provider>
        
        )
        expect(wrapper.find('.auth__alert-error').exists()).toBeTruthy;      
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initialState.ui.msgError)


    })
    
    


})
