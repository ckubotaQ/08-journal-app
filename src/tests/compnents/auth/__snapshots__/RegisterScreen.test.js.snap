// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreeb /> Debe de mostarse correctamente 1`] = `
<div
  className="animate__animated animate__bounceIn"
>
  <h1
    className="auth__title"
  >
    Register
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Carlos"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email"
      type="text"
      value="ckubota54@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Alredy registered?
    </a>
  </form>
</div>
`;
